### PR TITLE
Notebook launch errors

### PR DESCRIPTION
**Pull request by ciyer**
# Description

This should better handle the case that sometimes happens where a user tries to start a notebook and is shown an endless spinning wheel.

The source of the problem is a 500 response from the notebook server when the UI asks to start a notebook.

# Behavior

If the server returns an error on notebook launch, it is retried. If the second attempt also fails, an info message is shown:

image

## Testing

This is hard to test because it does not happen very frequently. To simulate, you can change the implementation of the function in notebook-servers.js:75 to:
```
  client.startNotebook = async (namespacePath, projectPath, branchName, commitId, options) => {
    const headers = client.getBasicHeaders();
    headers.append("Content-Type", "application/json");
    const url = `${client.baseUrl}/notebooks/servers`;
    const parameters = {
      namespace: decodeURIComponent(namespacePath),
      project: projectPath,
      commit_sha: commitId,
      branch: branchName,
      ...options
    };

    return new Promise((resolve, reject) => {
      setTimeout(() => {
        reject(new Error("testing"));
      }, 3000)
    });

    // return client.clientFetch(url, {
    //   method: "POST",
    //   headers,
    //   body: JSON.stringify(parameters)
    // }).then((resp) => {
    //   return resp.data;
    // });
  };
```

### CODE REVIEW

**Comment by ciyer** [comment]
> If I correctly understood the problem, the environment usually starts despite the `500` error, is this correct?

No, the server does not start. Since the exception was discarded before, the UI would be forever stuck "starting server" spinning wheel state, and refreshing the browser would show that no server had started, so you would have to try to launch a server again (and it would then usually work). 

> This results in the user having no feedback and being back to the "New environment" page.
> Wouldn't it be better to trigger a short timeout (5 seconds?) and then re-direct to the "Environment list" page?

I agree that it would be better to provide some feedback, but I think it would make sense to have a general user notification framework for non-fatal errors in place (there is a TODO in the code to add a user notification here). But in my experience, this has always worked the second time, so I feel like just showing the user the form again isn't such a bad choice.

**Comment by lorenzo-cavazzi** [comment]
> No, the server does not start. [...]

Oh my bad, I misunderstood the initial problem.

**Comment by rokroskar** [comment]
> The source of the problem is a 500 response from the notebook server when the UI asks to start a notebook.

do you mean "notebook server" or "notebook service"?

**Comment by ciyer** [comment]
> > The source of the problem is a 500 response from the notebook server when the UI asks to start a notebook.
> 
> do you mean "notebook server" or "notebook service"?

I mean "notebook service".

**Comment by rokroskar** [comment]
Do you know which request fails?

**Comment by ciyer** [comment]
> Do you know which request fails?

It is `api/notebooks/servers`

Here is the log from the gateway:

```
10.42.48.3 - - [24/Jul/2020:20:38:07 +0000] "POST /api/notebooks/servers HTTP/1.1" 500 290 "-" "-" 8487 "notebooks@file" "https://sekhar.dev.renku.ch/jupyterhub" 1809ms
```

This is the log from the notebooks service:

```
[2020-07-24 20:38:07,697] DEBUG in notebooks: Request to create server: new-new-test-b91a2ca7 with options: {'namespace': 'cramakri', 'project': 'new-new-test', 'commit_sha': 'd94e7f1142bc8e767a38412273d12288b4861ba4', 'branch': 'master', 'serverOptions': {'cpu_request': 0.1, 'defaultUrl': '/lab', 'gpu_request': 0, 'lfs_auto_fetch': False, 'mem_request': '1G'}} for user: {'kind': 'user', 'name': 'cramakri', 'admin': False, 'groups': [], 'server': None, 'pending': None, 'created': '2020-07-23T08:58:36.859172Z', 'last_activity': '2020-07-24T20:37:25.932074Z', 'servers': None}
7/24/2020 10:38:08 PM [2020-07-24 20:38:08,881] INFO in gitlab_: Using image registry.dev.renku.ch/cramakri/new-new-test:d94e7f1.
7/24/2020 10:38:08 PM [2020-07-24 20:38:08,882] DEBUG in notebooks: Creating server new-new-test-b91a2ca7 with {'namespace': 'cramakri', 'project': 'new-new-test', 'branch': 'master', 'commit_sha': 'd94e7f1142bc8e767a38412273d12288b4861ba4', 'project_id': 600, 'notebook': None, 'image': 'registry.dev.renku.ch/cramakri/new-new-test:d94e7f1', 'git_clone_image': 'renku/git-clone:0.7.5-3096343', 'server_options': {'cpu_request': 0.1, 'defaultUrl': '/lab', 'gpu_request': 0, 'lfs_auto_fetch': False, 'mem_request': '1G'}}
7/24/2020 10:38:09 PM [2020-07-24 20:38:09,384] DEBUG in notebooks: {}
7/24/2020 10:38:09 PM [2020-07-24 20:38:09,414] ERROR in notebooks: creating server new-new-test-b91a2ca7 failed with 500
```

**Review by lorenzo-cavazzi — COMMENTED** [review]
Very elegant solution! It's working as advertised :+1: 

I added a couple of comments to slightly modify the output/content of the error message, feel free to disagree with the suggested changes :smile:

File: src/notebooks/Notebooks.container.js [hunk 1]
```diff
@@ -324,18 +325,33 @@ class StartNotebookServer extends Component {
...
+    //* Data from notebooks/servers endpoint needs some time to update properly.
+    //* To avoid flickering UI, just set a temporary state and display a loading wheel.
+    this.setState({ "starting": true });
+    this.internalStartServer().catch((error) => {
+      // Some failures just go away. Try again to see if it works the second time.
+      setTimeout(() => {
+        this.internalStartServer().catch((error) => {
+          // TODO: #991
+          // eslint-disable-next-line no-console
+          console.log(["Could not start server", error]);
```
[the remark this answers is not in the document]
Comment (ciyer) [reply src/notebooks/Notebooks.container.js:348]: This is intended. I'm not sure the error message from the server will be helpful to the user, but by putting it in the console, we can access it for debugging purposes. Later, when #991 is implemented, we can have some handling for this situation.

File: src/notebooks/Notebooks.present.js [hunk 2]
```diff
@@ -1204,7 +1204,7 @@ class StartNotebookServerOptions extends Component {
 
     return renderedServerOptions.length ?
       renderedServerOptions.concat(globalWarning) :
-      <label>Notebook options not avilable</label>;
+      <label>Notebook options not available</label>;
```
[the remark this answers is not in the document]
Comment (ciyer) [reply src/notebooks/Notebooks.present.js:1207]: I think that's a good idea. We can make an issue for this.

File: src/notebooks/Notebooks.present.js [hunk 3]
```diff
@@ -1300,12 +1300,25 @@ class ServerOptionLaunch extends Component {
...
     const globalNotification = (warnings.length < 1) ?
       null :
       <Warning key="globalNotification">
         The environment cannot be configured exactly as requested for this project.
         You can still start one, but some things may not work correctly.
       </Warning>;
+    const launchErrorAlert = (launchError != null) ?
+      <Fragment key="launch-error">
+        <br />
+        <InfoAlert timeout={0}>
```
[the remark this answers is not in the document]
Comment (ciyer) [reply src/notebooks/Notebooks.present.js:1313]: I agree that it's not obvious how best to handle this. In this case, every time I have seen an error, just clicking "Start Server" again works. So, for that case, I think "Info" is not a bad way to go. But there may be other errors that we haven't yet seen that would better be displayed as Warn or Error. For that reason, I'm in favor of leaving as is and revisiting in the future if we encounter different kinds of errors here.

File: src/notebooks/Notebooks.present.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (lorenzo-cavazzi) [reply src/notebooks/Notebooks.present.js:1207]: https://github.com/SwissDataScienceCenter/renku-ui/issues/1019

File: src/notebooks/Notebooks.container.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (lorenzo-cavazzi) [reply src/notebooks/Notebooks.container.js:348]: #991 will show the error to the user anyway. In general, I'm not sure having a `console.log` is that useful, I would rather send the error to sentry. But we can leave it like it is now

**Review by lorenzo-cavazzi — APPROVED** [review]
I still don't like much the `console.log` solution (see the comment inline) but if you think it's fine I'd say to go on and merge this :+1:

**Pull request merged**